### ISSUE TITLE
33 words for 256 bit shamir backup?

### ISSUE THREAD

**Issue by vincecate** [seedhammer/seedhammer#17]
With seedhammer we should all move to 256 bit seeds, no excuse for 128 bit seeds and 12 words
once you have automatic engraving.

Shamir Backup is also easy with automatic engraving.   In the past engraving 7 plates of 33 words each
is just too much work.   But with seedhammer this would not be bad at all.

It looks like we could fit the 33 words on one side.  It seems 3 columns of 11 words each could fit.
Might have to remove the numbers or make them smaller.  The QR code could go on the back.  

Can we get support for 33 words?   My son and I might be able to modify the code ourselves but we might
also be willing to fund a developer to do so.   It would be nice if it could just scan a QR code with 33 words
and engrave the 33 words.

I have my own code to make new 256 seeds and generate the shamir shares and QR codes for these.
Will put this on github in awhile.  Just about 200 lines of python.  So someone really can check it over on their own.

**Comment by seedhammer** [comment]
We are certainly interested in adding support for Shamir's Shared Secret seeds. I imagine the user interface for entering a SLIP39 share would be similar to how you enter a passphrase protected seed for a singlesig.

The issue we see is that we strongly prefer keeping the backside of the plates for the descriptor for easy identification of the wallet from any quorum of plates (in this case any 1 plate). So for 256-bit SLIP39 to be feasible, we'll have to fit 33 words plus their QR representation on one side.

**Comment by seedhammer** [comment]
Looking closer at the [SLIP39](https://github.com/satoshilabs/slips/blob/master/slip-0039.md#format-of-the-share-mnemonic) standard, it seems somewhat wasteful compared to BIP39:

- Spends 15 bits on the identifier, BIP39 has no identifier
- Spends 30 bits on checksum, BIP39 8 bits.
- Wordlist is 1024 long which amounts to 10 bits per word, BIP39's wordlist is 2048 long which amounts to 11 bits per word.

In summary, 256 bit security takes 24 BIP39 words and 33 SLIP39 words.

Perhaps most impractical is that SLIP39 is not widespread and doesn't have a standard QR representation because Trezor doesn't have a camera.

**Comment by vincecate** [comment]
The Keystone Pro wallet also supports SLIP39 and it has a camera but it
does not currently have the ability to load a seed from a QR code.
Entering seeds into it is not that painful though.  It has a QWERTY
touchscreen and after 2 or 3 letters it will list the word and you just
touch it.
So it goes fast, even for 33 words.

While the QR code may not be used now, I expect I may not need the
stainless steel plates for 20 years and by that time I expect wallets can
scan in a QR code with a list of words.

I use Electrum and it does not need a descriptor to recover a wallet.   It
does seem like storing something about the type of wallet would be  a
good thing but the descriptor seems long and complicated.   Not easy to
enter correctly.   Could Electrum or other wallets even use it?

My seed maker code can output QR codes, so if seed hammer can scan them in
that makes it easy to do the engraving,
even if SLIP39 ends up being a lot of words.  Can it scan in my QR code?
I have attached a sample share QR code.

   -- Vince

**Comment by vincecate** [comment]
Attached is my seed maker code so far (maybe not going to be official
name).   I will have a couple friends check it over before putting it on
github but just in case you wanted to see it.

  -- Vince

**Comment by seedhammer** [comment]
Apart from the 33 word issue, I think another important issue is standardization. #14 is a request for [SSKR](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-011-sskr.md) support, and recently [codex32](https://github.com/bitcoin/bips/blob/master/bip-0093.mediawiki#user-content-Abstract) is gaining traction, totalling 3 standards (slip39, codex32/bip93, sskr), only one with an official QR encoding.

Ideally, we'd like to support just a single Shamir's scheme that we can then help push to signing devices as a standard.

This is obviously a longer term issue, so if you want something that covers your usecase right now, and already have a computer with your SLIP39 generator/splitter, I suggest looking into our [CLI](https://github.com/seedhammer/seedhammer/tree/main/cmd/cli). It's not an officially supported tool, but it does offer a more immediate interface with the SeedHammer compared to the controller software.

**Comment by seedhammer** [comment]
> only one with a QR encoding [SSKR]

This is not entirely true: the [codex32](https://github.com/bitcoin/bips/blob/master/bip-0093.mediawiki#user-content-Abstract) standard encodes shares in bech32 with a distinguished header, "ms1". So we could simply choose to encode a codex32 part to QR as its bech32 encoding, which makes it exactly match the human readable counterpart.

**Comment by vincecate** [comment]
> Apart from the 33 word issue, I think another important issue is
> standardization. #14 <https://github.com/seedhammer/seedhammer/issues/14>
> is a request for SSKR
> <https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-011-sskr.md>
> support, and recently codex32
> <https://github.com/bitcoin/bips/blob/master/bip-0093.mediawiki#user-content-Abstract>
> is gaining traction, totalling 3 standards, only one with a QR encoding.
>
> Ideally, we'd like to support just a single Shamir's scheme that we can
> then help push to signing devices as a standard.
>
Sure.   I think I would be perfectly happy with SSKR.   I think codex32
looks like far too much work for the lazy humans I know.
We like to use machines to do grunt work and calculations for us.

I also think that by the time I need to use these stainless steel plates
that hardware wallets then will be able to use the
camera and OCR to read in regular words, so that QR code is not really
necessary in the future.

> This is obviously a longer term issue, so if you want something that
> covers your usecase right now, and already have a computer with your SLIP39
> generator/splitter, I suggest looking into our CLI
> <https://github.com/seedhammer/seedhammer/tree/main/cmd/cli>. It's not an
> officially supported tool, but it does offer a more immediate interface
> with the SeedHammer compared to the controller software.
>
Thanks.   That does look like a reasonable possibility.

  -- Vince

**Comment by seedhammer** [comment]
> Sure. I think I would be perfectly happy with SSKR. I think codex32 looks like far too much work for the lazy humans I know. We like to use machines to do grunt work and calculations for us.

Codex32 allows you to generate shards completely by hand, but it's not required. I assume you're using a program to generate your slip39 shards as well?
